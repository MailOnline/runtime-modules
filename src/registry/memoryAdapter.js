import {compare} from 'semver';

const defaultRegistry = new Map();

const add = ({name, version, registry = defaultRegistry}) => {
  const current = registry.get(name) || {
    name,
    versions: []
  };

  if (current.versions.includes(version)) {
    throw new Error('Version already in registry');
  }

  const newVersions = current.versions.slice();

  newVersions.push(version);
  newVersions.sort(compare);

  registry.set(
    name,
    {
      ...current,
      versions: newVersions
    }
  );
};

const get = ({name, registry = defaultRegistry}) => registry.get(name) || null;

const list = ({registry = defaultRegistry} = {}) => {
  const values = [];

  registry.forEach((value) => {
    values.push(value);
  });

  return values;
};

export {
  add,
  get,
  list
};

const memoryAdapter = {
  add,
  get,
  list
};

export default memoryAdapter;
