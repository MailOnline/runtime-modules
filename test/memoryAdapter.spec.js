import {add, get, list} from '../src/registry/memoryAdapter';

describe('memoryAdapter', () => {
  it('should add and sort multiple versions', () => {
    const registry = new Map();
    const firstPackage = '@scope/package-name';
    const secondPackage = 'other-package';
    const versionsToAdd = ['3.2.1', '1.2.3', '1.3.1'];
    const expectedVersions = ['1.2.3', '1.3.1', '3.2.1'];

    [firstPackage, secondPackage].forEach((name) => {
      // eslint-disable-next-line max-nested-callbacks
      versionsToAdd.forEach((version) => {
        add({
          name,
          registry,
          version
        });
      });
    });

    [firstPackage, secondPackage].forEach((name) => {
      expect(registry.get(name).versions).toEqual(expectedVersions);
    });
  });

  it('should get package by name', () => {
    const registry = new Map();
    const name = '@scope/package-name';
    const versions = ['1.2.3', '1.3.1', '3.2.1'];

    versions.forEach((version) => {
      add({
        name,
        registry,
        version
      });
    });

    const packageInfo = get({
      name,
      registry
    });

    expect(packageInfo).toMatchSnapshot();
  });

  it('should list packages in the registry', () => {
    const registry = new Map();
    const firstPackage = '@scope/package-name';
    const secondPackage = 'other-package';
    const versionsToAdd = ['3.2.1', '1.2.3', '1.3.1'];

    [firstPackage, secondPackage].forEach((name) => {
      // eslint-disable-next-line max-nested-callbacks
      versionsToAdd.forEach((version) => {
        add({
          name,
          registry,
          version
        });
      });
    });

    expect(list({registry})).toMatchSnapshot();
  });
});
