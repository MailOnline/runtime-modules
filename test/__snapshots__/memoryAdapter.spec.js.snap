// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`memoryAdapter should get package by name 1`] = `
Object {
  "name": "@scope/package-name",
  "versions": Array [
    "1.2.3",
    "1.3.1",
    "3.2.1",
  ],
}
`;

exports[`memoryAdapter should list packages in the registry 1`] = `
Array [
  Object {
    "name": "@scope/package-name",
    "versions": Array [
      "1.2.3",
      "1.3.1",
      "3.2.1",
    ],
  },
  Object {
    "name": "other-package",
    "versions": Array [
      "1.2.3",
      "1.3.1",
      "3.2.1",
    ],
  },
]
`;
